node/src/components/storage.rs

```diff
@@ -1212,6 +1234,49 @@ fn check_block_metadata_db(
             ),
         }
     }
-    info!("Check for block_metadata_db complete");
+
+    drop(cursor);
+    txn.commit()?;
+
+    info!("block metadata database initialized");
+    Ok(())
+}
+
+/// Purges stale entries from the deploy metadata database.
+fn initialize_deploy_metadata_db(
+    env: &Environment,
+    deploy_metadata_db: &Database,
+    deleted_block_hashes: &HashSet<BlockHash>,
+) -> Result<(), LmdbExtError> {
+    info!("initializing deploy metadata database");
+    let mut txn = env.begin_rw_txn()?;
+    let mut cursor = txn.open_rw_cursor(*deploy_metadata_db)?;
+
+    for (raw_key, raw_val) in cursor.iter() {
+        let mut deploy_metadata: DeployMetadata = lmdb_ext::deserialize(raw_val)?;
+
+        let mut removed = false;
+        for block_hash in deleted_block_hashes {
+            if deploy_metadata
+                .execution_results
+                .remove(block_hash)
+                .is_some()
+            {
+                removed = true;
+            }
+        }
+
+        if deploy_metadata.execution_results.is_empty() {
+            cursor.del(WriteFlags::empty())?;
+        } else if removed {
+            let buffer = lmdb_ext::serialize(&deploy_metadata)?;
+            cursor.put(&raw_key, &buffer, WriteFlags::empty())?;
```

Comment:
Can you add a comment that we remove the whole entry under the cursor, from the DB, if there are no more blocks it is contained in?